### PR TITLE
Range filters should be improved to work mostly client side

### PR DESCRIPTION
**Pull request by mgbaybay**
## Changes
- Range filters should be improved to work mostly client side
- Reports changes

Relates to: https://github.com/xibosignage/xibo/issues/2732

### CODE REVIEW

File: reports/apirequests-report-form.twig [hunk 1]
```diff
@@ -44,11 +44,14 @@
                 <div class="XiboFilterCustom">
                     <div class="FilterDiv card-body" id="apiRequestsHistoryFilter">
                         <form class="form-inline">
+                            {% set title %}{% trans "Range" %}{% endset %}
+                            {{ inline.dateRangeFilter("reportFilter", title, "", "", "", "", "") }}
```
Comment (dasgarner) [inline reports/apirequests-report-form.twig:48]: I think the date range from/to date fields should be included in the dateRangeFilter macro automatically, so you do not need to add them separately in the form.

That would mean we can name them accordingly and have more than one range filter per form if we wanted to.

**Review by dasgarner — CHANGES_REQUESTED** [review]
This is going in the right direction, although I have a few suggestions/comments to look at please.

File: lib/Report/ApiRequests.php [hunk 2]
```diff
@@ -174,57 +174,18 @@ public function getResults(SanitizerInterface $sanitizedParams)
             throw new AccessDeniedException();
         }
 
-        $currentDate = Carbon::now()->startOfDay();
-
         //
         // From and To Date Selection
```
[the remark this answers is not in the document]
Comment (mgbaybay) [reply lib/Report/ApiRequests.php:178]: Thank you! Updated the comment for the reports [here](https://github.com/xibosignage/xibo-cms/pull/2827/commits/8fec6549831d9099a87d0bb3101f9a7c9e6ae736).

File: lib/Report/ApiRequests.php [hunk 3]
```diff
@@ -174,57 +174,18 @@ public function getResults(SanitizerInterface $sanitizedParams)
...
-                $fromDt = $fromDt->startOfDay();
-
-                // If toDt is current date then make it current datetime
-                if ($toDt->format('Y-m-d') == $currentDate->format('Y-m-d')) {
-                    $toDt = Carbon::now();
-                } else {
-                    $toDt = $toDt->addDay()->startOfDay();
-                }
+        // Range
+        $reportFilter = $sanitizedParams->getString('reportFilter');
```
[the remark this answers is not in the document]
Comment (mgbaybay) [reply lib/Report/ApiRequests.php]: Thanks for this! Removed unused variables [here](https://github.com/xibosignage/xibo-cms/pull/2827/commits/8fec6549831d9099a87d0bb3101f9a7c9e6ae736).

File: views/forms.twig [hunk 4]
```diff
@@ -393,3 +393,20 @@
         </div>
     </div>
 {% endmacro %}
+
+{% macro dateRangeFilter(name, title, value, helpText, groupClass, validation, accessKey) %}
+    <div class="form-group mr-1 mb-1 {{ groupClass }}">
+        <label class="control-label mr-1" title="{{ helpText }}" for="{{ name }}" accesskey="{{ accessKey }}">{{ title }}</label>
+        <select class="form-control XiboDateRangeFilter" name="{{ name }}" id="XiboDateRangeFilter">
+            <option value="" >Select a range</option>
```
[the remark this answers is not in the document]
Comment (mgbaybay) [reply views/forms.twig]: Thanks, Dan! Translation has been added [here](https://github.com/xibosignage/xibo-cms/pull/2827/commits/8fec6549831d9099a87d0bb3101f9a7c9e6ae736).

File: reports/apirequests-report-form.twig (hunk 1 above)
Comment (mgbaybay) [reply reports/apirequests-report-form.twig:48]: Thanks again, Dan! Refactored the code to simply the input handling [here](https://github.com/xibosignage/xibo-cms/pull/2827/commits/8361006629f23de084429d81c2938931a8714076).

File: views/forms.twig [hunk 5]
```diff
@@ -393,3 +393,20 @@
         </div>
     </div>
 {% endmacro %}
+
+{% macro dateRangeFilter(name, title, value, helpText, groupClass, validation, accessKey) %}
+    <div class="form-group mr-1 mb-1 {{ groupClass }}">
+        <label class="control-label mr-1" title="{{ helpText }}" for="{{ name }}" accesskey="{{ accessKey }}">{{ title }}</label>
+        <select class="form-control XiboDateRangeFilter" name="{{ name }}" id="XiboDateRangeFilter">
```
[the remark this answers is not in the document]
Comment (mgbaybay) [reply views/forms.twig]: Thank you! Removed the ID [here](https://github.com/xibosignage/xibo-cms/pull/2827/commits/8361006629f23de084429d81c2938931a8714076).

File: views/forms.twig [hunk 6]
```diff
@@ -393,3 +393,65 @@
...
+        <div class="form-group hidden mr-1 rangeFilterInput">
+            <label class="control-label mr-1" title="{{ helpText }}" for="{{ name }}" accesskey="{{ accessKey }}">
+                {% trans "From Date" %}
+            </label>
+            <div class="input-group">
+                <div class="input-group-prepend input-group-text date-open-button" role="button">
+                    <i class="fa fa-calendar"></i>
+                </div>
+                <input class="form-control dateControl date rangeInput"
+                       type="text" name="fromDt" id="fromDt"
```
Comment (dasgarner) [inline views/forms.twig:427]: Great, this reusable control is much better! Do we need to think about making the IDs here be specific to this control? If we include this macro more than once we will currently have duplicated IDs. The name would need to stay the same to it matches the API, but the ID could be unique to this instance (perhaps we could make the ID fromDt_name?)

**Review by dasgarner — CHANGES_REQUESTED** [review]
I have one more comment here, and I think @maurofmferrao needs to review the JS please.

File: views/forms.twig (hunk 6 above)
Comment (mgbaybay) [reply views/forms.twig]: Hi, Dan! Thank you for this feedback! Please see updated changes [here](https://github.com/xibosignage/xibo-cms/pull/2827/commits/ec9e3dad2a216a20738eb41dab70a5513515fa50) :)

File: ui/src/core/xibo-cms.js [hunk 7]
```diff
@@ -109,6 +109,14 @@ window.XiboInitialise = function(scope, options) {
     scope = ' ';
   }
 
+  // Search for any grids on the page and render them
+  $(scope + " .XiboGrid").each(function() {
+    // Custom Date Range Filter
+    $(this).find('select.XiboDateRangeFilter').on('change', function() {
```
Comment (maurofmferrao) [inline ui/src/core/xibo-cms.js:115]: With the new refactor for all the JS files, EsLint complains with "Disallow use of this in contexts where the value of this is undefined" in this case (https://eslint.org/docs/latest/rules/no-invalid-this).

Can you please use something like:

```
$(scope + " .XiboGrid").each(function(_idx, el) {
  // Custom Date Range Filter
  $(el).find('select.XiboDateRangeFilter').on('change', function() {
  (...)
```

File: ui/src/core/xibo-cms.js [hunk 8]
```diff
@@ -109,6 +109,14 @@ window.XiboInitialise = function(scope, options) {
     scope = ' ';
   }
 
+  // Search for any grids on the page and render them
+  $(scope + " .XiboGrid").each(function() {
```
Comment (maurofmferrao) [inline ui/src/core/xibo-cms.js:113]: Can you replace the double quotes with single quotes here please (https://eslint.org/docs/latest/rules/quotes)?

File: ui/src/core/xibo-cms.js [hunk 9]
```diff
@@ -3079,6 +3087,103 @@ function createColorPicker(element, options) {
...
+
+/**
+ * Create custom date range filter
+ * @param {object} element jquery object or CSS selector
+ */
+function updateDateRangeFilter(element) {
+  const selected = $(element).find('option:selected').val();
+  const $form = $(element).closest('form');
+
+  // This would allow us to target this specific element in case there are multiple range filters used
```
Comment (maurofmferrao) [inline ui/src/core/xibo-cms.js:3099]: All looks good in this function, it's just this line that has more than 80 char of lenght (https://eslint.org/docs/latest/rules/max-len). Can you please split it into 2 comments?

File: ui/src/core/xibo-cms.js [hunk 10]
```diff
@@ -3079,6 +3087,103 @@ function createColorPicker(element, options) {
...
+ * Calculates from/to date range selection
+ * @param {object} $form jquery object or CSS selector
+ * @param {string} selectFilterId filter ID selector
+ */
+function updateDateRangeSelection($form, selectFilterId) {
+  let fromDt = moment().startOf('day');
+  let toDt = moment().endOf('day');
+
+  $(`.rangeFilterInput_${selectFilterId}`).on('change', function() {
+    fromDt = moment($(`#fromDt_${selectFilterId}`).val()).startOf('day').format();
```
Comment (maurofmferrao) [inline ui/src/core/xibo-cms.js:3161]: Same thing as the function above (https://eslint.org/docs/latest/rules/max-len).

**Review by maurofmferrao — CHANGES_REQUESTED** [review]
Just some minor changes regarding the code formatting, everything else looks great! 👍🏻

File: ui/src/core/xibo-cms.js (hunk 7 above)
Comment (mgbaybay) [reply ui/src/core/xibo-cms.js]: Thank you, Mauro! Please see changes [here](https://github.com/xibosignage/xibo-cms/pull/2827/commits/b972050eb3cdc138c24aeb1e1e98b4b6f6df41b0) :)

File: ui/src/core/xibo-cms.js [hunk 11]
```diff
@@ -3079,6 +3087,112 @@ function createColorPicker(element, options) {
...
+ * @param {object} $form jquery object or CSS selector
+ * @param {string} selectFilterId filter ID selector
+ */
+function updateDateRangeSelection($form, selectFilterId) {
+  let fromDt = moment().startOf('day');
+  let toDt = moment().endOf('day');
+
+  $(`.rangeFilterInput_${selectFilterId}`).on('change', function() {
+    fromDt = moment($(`#fromDt_${selectFilterId}`).val())
+            .startOf('day')
```
Comment (maurofmferrao) [inline ui/src/core/xibo-cms.js:3163]: This should be indented with only 2 more spaces than the line above.
( same for `toDt` and for the `val` + `format` in `formatInputFields` below )

**Review by maurofmferrao — CHANGES_REQUESTED** [review]
Some more code formatting changes. 🙏🏻

File: ui/src/core/xibo-cms.js [hunk 12]
```diff
@@ -109,6 +109,14 @@ window.XiboInitialise = function(scope, options) {
     scope = ' ';
   }
 
+  // Search for any grids on the page and render them
+  $(scope + ' .XiboGrid').each(function(_idx, el) {
+    // Custom Date Range Filter
+    $(el).find('select.XiboDateRangeFilter').on('change', function() {
+      updateDateRangeFilter(this);
```
[the remark this answers is not in the document]
Comment (mgbaybay) [reply ui/src/core/xibo-cms.js]: Thank you again, Mauro! Please see updates [here](https://github.com/xibosignage/xibo-cms/pull/2827/commits/8606ed29244156017d1b97c823bc6ed79575c04e) :)

**Review by maurofmferrao — APPROVED** [review]
Looks good to me! ✅

**Comment by dasgarner** [comment]
@maurofmferrao how do you recommend this PR is tested? should the test wait until we've retargeted it to `develop`?

**Comment by maurofmferrao** [comment]
> @maurofmferrao how do you recommend this PR is tested? should the test wait until we've retargeted it to `develop`?

It might work for the `new develop`, and we can test now for the current `kopff` just to make sure it works. But maybe to avoid retesting we do that ( wait for 4.2 to become `develop` and then test+merge )

**Comment by ifarzana** [comment]
TEST: /Reporting END

**Comment by ifarzana** [comment]
TEST: /Reporting/report_timeconnectedsummary.cy.js END

**Comment by ifarzana** [comment]
The cypress test of Time Connected Summary report needs to be updated to reflect the new changes. I am signing off on this work for now, and we will address the failed test later.

**Comment by ifarzana** [comment]
TEST: login.cy.js END

**Review by maurofmferrao — APPROVED** [review]
Looks good to me!

**Pull request merged**